# <font color='steelblue'>270. Introducción a la modelización estadística</font>

**Material desarrollado por los [equipos de trabajo de IA4LEGOS](https://ia4legos.umh.es/)**


**Fecha última edición**: 10/05/2025

**Licencia**: <small><a rel="license" href="http://creativecommons.org/licenses/by-sa/4.0/"><img alt="Creative Commons License" style="border-width:0" src="https://i.creativecommons.org/l/by-sa/4.0/88x31.png" /></a><br /></small>

No olvides hacer una copia si deseas utilizarlo. Al usar estos contenidos, aceptas nuestros términos de uso y nuestra política de privacidad.

In [ ]:
#@title <font color="steelblue" size="+1"> Configuración del cuaderno
# @markdown Para garantizar la funcionalidad completa de este cuaderno, ejecuta esta celda de código que carga las librerías básicas de Python para el análisis de datos y su representación gráfica.

# Librerías python
##################

# De sistema
from io import StringIO
import sys
# Matemáticas
import numpy as np
import math
# Análisis de datos
import pandas as pd
# Gráficas
%matplotlib inline
import matplotlib.pyplot as plt
import seaborn as sns

# específicos de statsmodels
import statsmodels
import statsmodels.api as sm
import statsmodels.formula.api as smf

# específicos de Formulaic
import patsy
! pip install formulaic
import formulaic
from formulaic import model_matrix

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 115.7/115.7 kB 569.1 kB/s eta 0:00:00


# <font color="steelblue">Descripción del cuaderno</font>

**Descripción:** En este cuaderno se hace una introducción a la modelización estadística. Hablamos sobre variabilidad, poblaciones y muestras, parámetros y estimadores. Planteamos cuáles son los objetivos y retos de la modelización estadística, cuál el flujo de trabajo en la construcción de un modelo y comentamos los diversos tipos de modelos que trabajaremos durante el curso. Presentamos también en este cuaderno distintos bancos de datos que utilizaremos para aplicar lo que aprendamos. Introducimos las librerías de Python útiles en modelización estadística, y repasamos las bases del análisis exploratorio con Python sobre ejemplos concretos, para hacer patente cómo utilizarlo eficazmente antes de proceder con el ajuste de modelos.

**Recomendaciones antes de usarlo:** Conocimientos básicos del lenguaje Python, conceptos básicos de probabilidad y análisis exploratorio de datos.



# <font color='steelblue'> Contenidos

1. Introducción.
1. Conceptos básicos.
1. Modelos estadísticos y flujo de trabajo.
1. Retos en la modelización.
1. Librerías estadísticas en Python.
1. Análisis exploratorio con Python.
1. Bancos de datos.

# <font color="steelblue">1. Introducción</font>

La importancia de la Estadística dentro del campo experimental en todas las áreas de conocimiento es indiscutible: para extraer conclusiones de un conjunto de datos experimentales es necesaria la utilización de procedimientos estadísticos validados. Con los avances de la computación, de programas estadísticos diversos incluso para usuarios no especialistas en Estadística, así como del avance de la IA que estamos viviendo en los últimos años, la importancia de un conocimiento estadístico básico y una capacidad crítica bien desarrolladas, son realmente imprescincibles para abordar el análisis de la información a nuestro alrededor, de un modo riguroso y fiable.

Cuando un comportamiento observado en los datos se quiere extrapolar y generalizar a una población mayor, y así validar científicamente, es preciso acudir la **modelización estadística** y el análisis de sus resultados. Los modelos estadísticos nos permitirán aproximar y predecir matemáticamente el comportamiento de datos experimentales no observados, siempre con parsimonia y  cierto grado de incertidumbre. Evidentemente, la construcción de modelos estadísticos no es una tarea rutinaria y sencilla que deba tomarse a la ligera. La propia naturaleza de los datos observados puede dar una idea, al usuario familiarizado con la modelización estadística, de los posibles modelos útiles, pero el modelo final ajustado será el resultado de un proceso iterativo de construcción, verificación y validación, que puede resultar costoso en algunas situaciones.




# <font color="steelblue">2. Modelos estadísticos y flujo de trabajo</font>

## <font color='steelblue'> Elementos básicos

Puesto que ya controlamos la jerga básica estadística, repasemos la jerga específica que utilizaremos cuando emprendemos la modelización estadística.

* El modo óptimo de responder cuestiones de investigación o resolver problemas de interés pasa por plantear un diseño de experimentos. Sin embargo, como ya comentamos, no siempre será el caso, y en ocasiones utilizaremos datos secundarios que ya otros recopilaron y quizá con otros fines distintos a los nuestros. Este hecho no anula la fiabilidad de las conclusiones que se deriven con un modelo estadística, pero las limita por la representatividad de la muestra.

* Hablaremos de **muestra** como el conjunto de todos los datos recopilados sobre las variables de interés, para dar respuesta a las cuestiones planteadas a través de algún modelo estadístico. En la bibliografía también encontrarás algunas fuentes que denominan muestras a las observaciones individualizadas sobre todas las variables utilizadas.

* En modelización estadística habitualmente tenemos una característica o variable objetivo que tratamos de relacionar con otras que hemos observado, porque nos interesa especialmente predecirla en función de ellas. Se trata de la **variable respuesta**, también reconocida como **target** en los modelos de aprendizaje. En la bibliografía más tradicional se la reconoce como variable dependiente, por su dependencia de las que actuarán como predictoras. Si la respuesta es de tipo numérico utilizaremos modelos de regresión para buscar relaciones con el resto de variables. Si es de tipo categórico utilizaremos modelos de clasificación, pues el objetivo será clasificar a los sujetos respecto de tal variable, también denominada etiqueta en la jerga de aprendizaje automático.

* **Variables predictoras o explicativas** serán pues el conjunto de atributos o características disponibles sobre cada uno de los datos o registros de la muestra y con los que pretendemos predecir la variable respuesta. A menudo se representan, para cada registro, con un vector que contiene información sobre las características observadas. Estas variables pueden ser de tipo numérico o categórico y recibirán tratos distintos en los modelos que estudiaremos. En la jerga del Machine Learning, las variables predictoras se denominan **características**.

* Aludiremos con el término **parámetros** a aquellos descriptores de la población que pretendemos estimar, como por ejemplo la estatura media, o el tiempo medio de degradación, pero también la severidad de la relación lineal entre un predictor y una respuesta. Los modelos estadísticos plantearán fórmulas para relacionar los predictores con la respuesta a través de parámetros, y aproximarlos a través de los datos recopilados en la muestra.

* Surge después otro término que nos permite reajustar los modelos para conseguir mejores predicciones. Se trata de los **hiperparámetros** o parámetros libres vinculados a algunos modelos y que nos permiten plantear realmente modelos alternativos para las mismas variables y una misma formulación relacional.
Todos los algoritmos de ajuste de modelos utilizan unos valores por defecto para los hiperparámetros, si bien podemos manipularlos.

* Un concepto crítico en modelización tiene que ver con el criterio que se sigue para ajustar un modelo. El ajuste de un modelo está siempre basado en minimizar una **función de pérdida**, que mide la diferencia entre el valor de la respuesta predicha por el modelo estadístico y el valor observado.


* Finalmente, comentar que cualquier banco de datos con el que emprenderemos la modelización tendrá una **estructura matricial**, donde las filas identificarán los registros o sujetos disponibles, y las columnas representarán las variables medidas y a relacionar. El cruce de una fila y una columna contendrá el valor observado para el sujeto registrado en dicha fila, en la variable representada en la columna en cuestión.


La modelización estadística pretendrá buscar respuestas a objetivos o hipótesis sobre la caracterización de una población, aproximando  la relación entre las características implicadas e identificando qué condiciones generan respuestas distintas. El analista siempre buscará el mejor modelo que dé respuestas válidas a los objetivos planteados. Y ahí... el abanico de posibilidades es inmenso, como veremos.

## <font color="steelblue">Formulación</font>

Un modelo estadístico es realmente una fórmula matemática que nos va a permitir relacionar la variable respuesta con las predictoras, para predecir la primera a través de las segundas. A esta fórmula matemática determinista, llamada **componente sistemática** del modelo, se le añade el error, que representa la variabilidad intrínseca en la población (sujetos con las mismas caracterizaciones pueden no compartir una misma respuesta), la que no es explicable con la fórmula (que es una aproximación a la relación real), y el error en la medición (que es esperable que sea pequeño si el sistema de medición ha sido validado). Este error o desviación se conoce como **componente aleatoria**. Comencemos a expresar estas ideas matemáticamente.

Hablamos ya de variables predictoras, que vamos a identificar como $(X_1, X_2,X_3, …)$ y de una (o más) variable respuesta o *target* que identificaremos como $Y$, y cuya predicción denotaremos por $\hat{Y}$. La modelización estadística busca relacionar la predicción con las predictoras mediante una fórmula o modelo matemático que expresaremos con la componente sistemática:

\begin{equation}
  \hat{Y} = f(X_1,X_2,...,X_k)
\end{equation}

Si nuestro modelo es 'bueno', esta función debe reflejar el comportamiento medio esperado de la variable respuesta, esto es, aproximar bien $E(Y)$ con $\hat{Y}$. La desviación entre ambas la denotaremos por $\epsilon$, de modo que el modelo estadístico lo podemos formular como:

\begin{equation}
  Y = \hat{Y} + \epsilon = f(X_1,X_2,...,X_k) + \epsilon
\end{equation}

Por último, mencionar que el error $\epsilon$ en ocasiones puede tener una estructura más compleja que surgirá en modelos avanzados que estudiaremos también en este curso.


## <font color="steelblue">Tipos de modelos</font>

En función del tipo de variable respuesta, las predictoras, de la relación a establecer entre ellas a través de $f$, y del establecimiento de las estructuras para los errores, tendremos diferentes tipos de modelos. En este caso consideramos los modelos más sencillo en ciancia de datos:

* **Modelos de reregresión lineal**, cuando el target es de tipo continuo.
* **Modelos de reregresión logística**, cuando el target es de tipo discreto.


## <font color="steelblue">Flujo de trabajo</font>

Aunque cada problema de modelización estadística es único, siempre se sigue un flujo de trabajo similar. En esta sección, aprenderemos a abordar los problemas de modelización de forma sistemática, siguiendo los siguientes pasos:

1. Definir y formular el problema.
1. Recoger los datos.
1. Describir los datos con un análisis exploratorio (AED).
1. Preprocesar los datos.
1. Elegir y entrenar un modelo de partida.
1. Mejorar el modelo.
1. Evaluar y validar el modelo.
1. Implantar y automatizar el modelo.


### Evaluar y validar el modelo

Cuando se ha seleccionado un modelo que funciona, es habitual evaluarlo para valorar su rendimiento final en términos de capacidad explicativa, es decir, cómo son de buenas las características para expicar el comportamiento de la respuesta. En función del tipo de target (numérico o categórico) podemos definir diferentes métricas de validación del modelo. En este punto presentamos las más generales pero profundizaremos más en este tema conforme vayamos presentando los diferentes tipos de modelos.

Las métricas de evaluación se utilizan para medir el rendimiento de los modelos de aprendizaje automático. Al principio de esta introducción comentamos que la mayoría de los problemas se refieren a cuestiones resolubles por regresión o por clasificación; en función del objetivo y el tipo de respuesta, la evaluación del modelo será diferente. Veamos las alternativas.

#### Métricas para problemas de regresión

En las tareas de regresión el objetivo es predecir el valor numérico de la variable respuesta. Así pues, las métricas que surgen de manera natural para evaluar estos modelos son las basadas en distancias entre las observaciones y las predicciones.

La diferencia entre el valor real de la respuesta ($y$) y el valor predicho ($\hat{y}$) se llama **error de predicción**:

$$Error = y - \hat{y}$$

El cuadrado del error de predicción sobre los datos observados $\{y_1,y_2,..., y_n\}$, se llama **error cuadrático medio** (MSE o *Mean Squared Error*, en inglés), y se calcula como:

$$MSE = \frac{\sum_{i=1}^n (y_i - \hat{y}_i)^2}{n}$$

Otra métrica común es la **raíz cuadrada del error cuadrático medio** (RMSE). El RMSE es la métrica de regresión más utilizada.

$$RMSE = \sqrt{\frac{\sum_{i=1}^n (y_i - \hat{y}_i)^2}{n}}$$

Hay veces que se trabaja con conjuntos de datos que contienen valores atípicos. Una métrica adecuada para este tipo de conjuntos de datos es el **error medio absoluto** (MAE, *Mean Absolute Error*) que se calcula como:

$$MAE = \frac{\sum_{i=1}^n |y_i - \hat{y}_i|}{n}$$

Por último, tenemos el **error de porcentaje medio absoluto** (MAPE) que es una medida relativa que escala esencialmente la MAE para que se muestre en unidades de porcentaje en lugar de en unidades de la variable. Se calcula como:

$$MAPE = \frac{1}{n} \sum_{i=1}^n \left|\frac{y_i - \hat{y}_i}{y_i}\right|$$

#### Métricas para problemas de clasificación

En los problemas de clasificación el objetivo es predecir correctamente la categoría de clasificación de cada observación. Así pues, las métricas naturales para evaluar estos modelos estarán basadas en contabilizar las coincidencias entre la clasificación correcta y la conseguida o predicha con el modelo.

Para definir estos indicadores, partimos de identificar las clasificaciones posibles, cuestión que salvamos con una **matriz de confusión**, esto es, una tabla que muestra el número de predicciones correctas e incorrectas realizadas por un clasificador en todas las clases disponibles. En el caso de una respuesta binaria, 0=Negativo y 1=Positivo en relación al hecho de que se de determinada característica, las clasificaciones/predicciones posibles sobre el total de datos o registros disponibles son las que se muestran en la siguiente matriz:

|                     |       | **Valor** | **Predicho** |
|---------------------|-------|--------------------|--------------|
|                     |       |        **0**       |     **1**    |
| **Valor** | **0** |         TN         |      FP      |
| **Observado**       | **1** |         FN         |      TP      |

Se identifican pues, los siguientes números:



|Abrev.|Índice |Descripción|
|-----|-------------------------|----------------|
**TP**| *True Positive* o Verdaderos positivos|# positivos clasificados positivos (1,1)|
**FP**| *False Positive* o Falsos positivos|# negativos clasificados positivos (0,1)|
**TN**|*True Negative* o Verdaderos negativos|# negativos clasificados negativos (0,0)|
**FN**|*False Negative* o Falsos negativos|# positivos clasificados negativos (1,0)|


En base a estos números se definen las siguientes métricas de evaluación de los modelos de aprendizaje basados en clasificación:

- La **exactitud** (representado en la literatura por A, de *accuracy*) es la métrica más utilizada y muestra la capacidad del modelo para hacer predicciones correctas. Se calcula a partir de la proporción de observaciones clasificadas correctamente (TP + TN sobre el total), esto es,

$$Exactitud (A)= \frac{TP + TN}{TP + TN + FP + FN}.$$

>La exactitud es un indicador de la bondad del modelo para hacer predicciones correctas, por lo que sólo será útil cuando el conjunto de datos esté repartido de forma equilibrada entre las categorías de respuesta, o no nos interese la clasificación en una categoría concreta.

>Supongamos que hemos construido un clasificador de imágenes que distingue entre caballos y humanos, con 250 imágenes de caballos y otras 250 de humanos. Si el modelo predice correctamente 400 imágenes en total (caballos y humanos), su exactitud es de 0.8, o lo que es lo mismo, del 80% en términos porcentuales.


>Cuando tenemos un conjunto de datos sesgado o desequilibrado hacia alguna categoría, o simplemente nos importa más predecir una categoría (como podría ser el diagnóstico de un tumor maligno), necesitamos una perspectiva diferente sobre cómo evaluar el modelo. Por ejemplo, si tenemos 450 imágenes de caballos y 50 imágenes de humanos, hay una probabilidad del 90% (450/500) de que el caballo se prediga correctamente porque el conjunto de datos está dominado por los caballos. Y aunque la predicción sea nefasta para los humanos, la exactitud seguirá siendo buena.

>Surgen pues otras métricas que, en estas circunstancias, serán más útiles que la exactitud, como la precisión, el recuerdo y la puntuación F1.

- La **precisión** (representado habitualmente como P, de *precision*) representa el porcentaje de observaciones correctamente clasificadas, de entre las que se han clasificado como positivas.

$$Precisión (P)= \frac{TP}{TP + FP}.$$

>Por ejemplo, nos podría interesar especialmente saber, de las imágenes que se clasificaron como humanos, cuántas eran realmente de humanos.

- Por otro lado, el **recuerdo** (representado por R, de *recall*) muestra el porcentaje de las observaciones que siendo positivas se clasificaron correctamente, y nos resulta útil cuando nos interesa especialmente la clasificación correcta de una determinada clase.

$$Recuerdo (R)= \frac{TP}{TP+FN}.$$

>En el ejemplo de caballos y humanos, descubriríamos, con el recuerdo, de las imágenes que eran de humanos, qué porcentaje fueron correctamente clasificadas como humanos.

>Dado que existe cierta relación entre la precisión y la recuperación, y a menudo el aumento de una conlleva la disminución de la otra, se propuso otra métrica construida a partir de las dos, para conjugar las ventajas de ambas.

- La **puntuación F1 (o *F1 Score*)** es la media armónica de la precisión y la recuperación, y muestra lo bueno que es el modelo a la hora de clasificar todas las clases sin tener que equilibrar la precisión y la recuperación. Si la precisión o la recuperación son muy bajas, la puntuación F1 también lo será.

$$F1 = \frac{2 \cdot P \cdot R}{P + R}.$$

**Evaluación y validación: procedimiento**

Todas estas métricas, en problemas de regresión o clasificación, se evalúan sobre los datos utilizados en el ajuste (muestra de entrenamiento), para comparar las predicciones con las observaciones. Si además se ha dejado una muestra test fuera del entrenamiento, también esta se evalúa para juzgar la calidad del modelo a la hora de predecir observaciones que no ha utilizado en el ajuste. Con las métricas disponibles, se decide sobre la bondad del ajuste.

La validación del modelo consistirá en estudiar si este es estable, es decir, si no se ve afectado en demasía por la muestra de entrenamiento y se resuelve habitualmente con técnicas de **validación cruzada**. Esta técnica se basa en que se separan los datos en varios subconjuntos (de tamaños similares) que se van utilizando para testar la bondad del modelo cuando se entrena el modelo con el resto. Se obtiene una medida media de calidad, promediando las métricas obtenidas en los subconjuntos de validación, y una medida de variabilidad/estabilidad con sus desviaciones típicas y rango de variación.


### Implantar y automatizar

Cuando todos los pasos anteriores han ido bien y se ha alcanzado un modelo con buenos resultados, el siguiente paso será implementar el modelo para que los usuarios puedan empezar a hacer uso de él, retroalimentando con nuevos datos con los que obtener mejores predicciones y servicios.



# <font color="steelblue">4. Retos en la modelización estadística</font>

Al igual que cualquier otra tecnología, la construción de modelos estadísticos ofrece desafíos, relacionados tanto con la mejora de los datos como con la mejora de los modelos de ajuste.

No puede haber buenos modelos con malos datos, pero es un hecho que en el mundo real es raro encontrar "buenos datos" a la primera. Por lo general los datos presentan desorden, errores, incoherencias, y se convierte en todo un arte la labor de procesarlos (depurar, etiquetar, e incluso crear nuevas variables) para conseguir una base de datos apta para iniciar el ajuste de un modelo.
Por otro lado, una vez disponemos de una base de datos razonablemente buena, el ajuste del modelo no es una labor trivial. En principio se inicia la labor de elegir qué modelo o modelos son razonables para los objetivos y el tipo de datos. A continuación hay que ajustarlos y compararlos para tomar una decisión.

Ya en el ajuste se nos pueden presentar dos tipos de problemas: el infra-ajuste (*underfitting*) y el sobre-ajuste (*overfitting*). Veamos en qué consisten.

El **infra-ajuste** o infra-adaptación del modelo se produce cuando el modelo da malos resultados (en términos de métricas de error) con los datos de entrenamiento (por ejemplo, una clasificación correcta de sólo el 50% de la muestra). Un problema de infra-ajuste ocasiona un problema de sesgo en la predicción. Esto puede ser causado porque el modelo es demasiado simple para los datos de entrenamiento o los datos no contienen las características que se están tratando de predecir. Para reducir este tipo de problema se puede optar por:

* Ajustar modelos más complejos. Modelos más sencillos, como los modelos lineales, pueden dar resultados más simples que los más complejos, como los modelos aditivos.

* Añadir más datos de entrenamiento y seleccionar mejor las mejores características para predecir,o incluso proponer transformaciones de las variables.

* Aumentar la complejidad de los modelos,lo que se conoce como "reducir el nivel de regularización del modelo". Por ejemplo, aumentar la sensibilidad de los modelos aditivos para adaptar mejor las variaciones en el rango de variación de la respuesta.

El **sobre-ajuste** o sobre-adaptación se produce cuando un modelo funciona muy bien con los datos de entrenamiento, pero no con datos nuevos que el modelo no utilizó para ajustar. Esto repercute en un inflado de la varianza de la muestra de validación.


El sobre-ajuste se puede deber a que el modelo es demasiado complejo para los datos disponibles o el volumen de la muestra de entrenamiento es demasiado pequeño. Para aliviar este tipo de problema, podemos intentar:

* Usar modelos más sencillos, simplificar el modelo actual a través de los hiperparámetros de los que dependen, reducir el número de predictores incluidos, etc.

* Utilizar más datos de entrenamiento para el ajuste.

* Aumentar el nivel de regularización del modelo, o lo que es lo mismo, reducir la complejidad del modelo, penalizando a los modelos que generan un error excesivamente pequeño para la muestra de entrenamiento.

# <font color="steelblue">5. Librerías estadísticas en Python</font>

De entre todos los paquetes o librerías disponibles en Python, disponemos específicamente de útiles para la manipulación de datos, el análisis descriptivo o la construcción de modelos estadísticos, en las librerías siguientes:

* Tratamiento de datos: `numpy`, `pandas`.
* Análisis gráficos:  `matplotlib` y `seaborn`.

* Modelización:
  
  * **`patsy`**, es útil para describir modelos estadísticos (especialmente modelos lineales o modelos que tienen un componente lineal) y construir matrices de diseño. Está siendo sustituído por **`Formulaic`**, pues dejó de desarrollarse en 2021.

  * **`statsmodels`**, proporciona clases y funciones para la estimación de muchos modelos estadísticos, así como para realizar pruebas estadísticas y exploración de datos.




## <font color="steelblue">Paquetes patsy/Formulaic

El paquete de Python [`patsy`](https://pypi.org/project/patsy/) es útil para construir matrices de diseño en los modelos lineales. En 2021 dejó de ser desarrollado, y se recomienda la migración a [Formulaic](https://github.com/matthewwardrop/formulaic).

Por ejemplo, si tenemos alguna variable `y` y queremos hacer una regresión con otras variables $x1$, $x2$, y la interacción entre ellas, entonces simplemente escribimos:

```
import patsy
patsy.dmatrices("y ~ x1 + x2 + x1:x2", datos)
```

y `patsy` se encarga de construir las matrices adecuadas.


La versión de esta funcionalidad en `Formulaic` es también sencilla de implementar con:
```
!pip install formulaic
import formulaic
from formulaic import model_matrix
model_matrix('y ~x1 + x2 + x1:x2', datos)
```

Además estos paquetes:

* Permite especificar transformaciones de datos usando código Python arbitrario: en lugar de x, podríamos haber escrito log(x), (x > 0), o incluso log(x) si x > 1e-5, de lo contrario log(1e-5),
* Proporciona una gama de opciones convenientes para codificar variables categóricas, incluida la detección automática y la eliminación de redundancias.
* Sabe cómo aplicar "la misma" transformación utilizada en datos originales a datos nuevos, incluso para transformaciones complicadas como el centrado o la estandarización (crítico si desea utilizar su modelo para hacer predicciones),
* Tiene un modo incremental para manejar conjuntos de datos que son demasiado grandes para caber en la memoria al mismo tiempo.
* Proporciona un lenguaje para la especificación simbólica y legible por humanos de matrices de restricciones lineales.
* Cuenta con una API simple para la integración en paquetes estadísticos.




## <font color="steelblue">Paquete statsmodels</font>

El paquete [`statsmodels`](https://www.statsmodels.org/stable/index.html) de Python proporciona clases y funciones para la estimación de muchos modelos estadísticos, así como para realizar pruebas estadísticas y exploración de datos estadísticos. Los resultados están testados con paquetes estadísticos existentes (R, Stata o SAS), para garantizar que son correctos. El paquete se publica bajo la licencia BSD modificada (3 cláusulas) de código abierto.


El código siguiente os permite aceder a los recursos del paquete necesarios para trabajar con los modelos de este curso y con el paquete `patsy`.
```
import statsmodels
import statsmodels.api as sm
import statsmodels.formula.api as smf
```


La documentación de `statsmodels` para los modelos que vamos a trabajar está disponible en la [Guía de uso](https://www.statsmodels.org/stable/user-guide.html), y en particular en:

* Modelos lineales: [`regression.linear_model.OLS`](https://www.statsmodels.org/stable/generated/statsmodels.regression.linear_model.OLS.html).
* Modelos lineales mixtos [regression.mixed_linear_model.MixedLM](https://www.statsmodels.org/stable/generated/statsmodels.regression.mixed_linear_model.MixedLM.html).
* Modelos lineales generalizados: [genmod.generalized_linear_model.GLM](https://www.statsmodels.org/stable/generated/statsmodels.genmod.generalized_linear_model.GLM.html).

# <font color="steelblue">6. Ejemplos</font>

En este punto se detallan algunas de las bases de datos que se van a usar a lo largo de este curso. Se realiza una breve descripción de cada una, se identifica el objetivo principal que se persigue y se indica el número de muestras y variables contenidas en ellas, así como si existen valores perdidos. En todos los casos se proporciona un enlace a la fuente utilizada, y también al fichero *raw* disponible en *Github*  para la carga del banco de datos (generalmente en formato .csv). En algunos de los casos se proporciona también un enlace a los datos preprocesados, es decir, que ya incluyen la imputación de valores perdidos, estandarización y codificación. En cada tema añadimos conjuntos de datos específicos.

Distinguimos los conjuntos de datos relacionados con modelos de regresión y aquellos abordables con modelos de clasificación.

## <font color="steelblue">Problemas de regresión</font>

### Diabetes

En un estudio sobre la diabetes se consideraron diez variables basales: edad, sexo, índice de masa corporal, presión arterial media y seis mediciones de suero sanguíneo, para 442 pacientes diabéticos. La respuesta de interés es una medida cuantitativa de la progresión de la enfermedad un año después del instante basal. Cada una de estas 10 variables predictoras se ha centrado en la media y se ha escalado por la desviación típica (es decir, la suma de los cuadrados de cada columna suma 1).

**Características del banco de datos**:

* Variable respuesta: `disease_progression`
* Tipo de variable respuesta: numérica
* Tipo de problema que se quiere resolver: regresión
* Número de registros: 442
* Número de variables: 11


**Variables contenidas**:
* **age**: edad (en años)
* **sex**: sexo
* **bmi**: índice de masa corporal
* **bp**: promedio de la presión sanguínea
* **s1 tc**: colesterol sérico total
* **s2 ldl**: lipoproteínas de baja densidad
* **s3 hdl**: lipoproteínas de alta densidad
* **s4 tch**: colesterol total
* **s5 ltg**: registro del nivel de triglicéridos en suero
* **s6 glu**: nivel de azúcar en sangre


**Referencias bibliográficas**:

* Bradley Efron, Trevor Hastie, Iain Johnstone and Robert Tibshirani (2004) “Least Angle Regression,” Annals of Statistics (with discussion), 407-499.


**Ficheros carga**:

* El banco de datos se puede cargar directamente desde scikit-learn,con el comando `sklearn.datasets.load_diabetes`.

### Electricity

Una central eléctrica de ciclo combinado (CCPP) está compuesta por turbinas de gas (GT), turbinas de vapor (ST) y generadores de vapor de recuperación de calor. En una CCPP, la electricidad es generada por turbinas de gas y vapor, que se combinan en un ciclo, y se transfiere de una turbina a otra. Mientras que el Vacío se recolecta y tiene efecto sobre la Turbina de Vapor, las otras tres variables ambientales afectan el desempeño del GT. Este conjunto de de datos se ha recogido de una central eléctrica de ciclo combinado a lo largo de 6 años (2006-2011), cuando la central eléctrica estaba configurada para funcionar a plena carga. Las características consisten en variables ambientales promedio por hora Temperatura (`AT`), Presión ambiental (`AP`), Humedad relativa (`RH`) y Vacío de escape (`V`) para predecir la producción de energía eléctrica neta por hora (`PE`) de la planta.



**Características del banco de datos**:

* Variable respuesta: `PE`
* Tipo de variable respuesta: numérica
* Tipo de problema que se quiere resolver: regresión
* Valores perdidos: no
* Número de registros: 9568
* Número de variables: 5


**Variables contenidas**:
Se recogen la promedios por hora de:
* **AT**: temperatura (en grados centígrados)
* **V**: vacío de escape (en cm Hg)
* **AP**: presión ambiental (en milibares)
* **RH**: humedad relativa (en %)
* **PE**: producción de energía eléctrica neta (en MW)


**Referencias bibliográficas**:

* Pınar Tüfekci, Prediction of full load electrical power output of a base load operated combined cycle power plant using machine learning methods, International Journal of Electrical Power & Energy Systems, Volume 60, September 2014, Pages 126-140, ISSN 0142-0615.

* Heysem Kaya, Pınar Tüfekci , Sadık Fikret Gürgen: Local and Global Learning Methods for Predicting Power of a Combined Gas & Steam Turbine, Proceedings of the International Conference on Emerging Trends in Computer and Electronics Engineering ICETCEE 2012, pp. 13-18 (Mar. 2012, Dubai).

* UCi Machine learning repository: https://archive.ics.uci.edu/ml/datasets/combined+cycle+power+plant


**Ficheros carga**:

* Github original: [(.csv)](https://raw.githubusercontent.com/ia4legos/MachineLearning/main/data/mult_linear_reg.csv)

### HousingCA

En este ejemplo vamos a utilizar la base de datos `HousingCA`, que recoge la información sobre el censo viviendas de California en el año 1990. Se está interesado en predecir el valor medio de la vivienda (`median_house_value`) medido en dólares americanos en función de las predictoras.


**Características del banco de datos**:

* Variable respuesta: `median_house_value`
* Tipo de variable respuesta: numérica
* Tipo de problema que se quiere resolver: regresión
* Valores perdidos: sí
* Número de registros: 20640
* Número de variables: 10



**Variables contenidas**:

* **median_house_value**: valor medio de la vivienda en dólares.
* **longitude**: medida de la distancia al oeste de una casa; un valor más alto es más al oeste.
* **latitude**: medida de la distancia al norte de una casa; un valor más alto es más al norte.
* **housing_median_age**: mediana de edad de una vivienda dentro de una manzana; un número más bajo es un edificio más nuevo (en años)
* **total_rooms**: número total de habitaciones dentro de un bloque.
* **total_bedrooms**: número total de dormitorios dentro de un bloque.
* **population**: Número total de personas que residen en un bloque.
* **households**: Número total de hogares, grupo de personas que residen en una unidad de vivienda, para un bloque.
* **median_income**: Ingreso medio de los hogares dentro de un bloque de casas (medido en decenas de miles de dólares estadounidenses)
* **oceanProximity**: Ubicación de la casa con respecto al océano/mar (`<1H OCEAN`, `INLAND`, `ISLAND`, `NEAR BAY`, `NEAR OCEAN`).


**Referencias bibliográficas**:

* Pace, R. Kelley and Ronald Barry, Sparse Spatial Autoregressions, Statistics and Probability Letters, 33 (1997) 291-297
* Kaggle: [enlace](https://www.kaggle.com/datasets/camnugent/california-housing-prices)


**Ficheros carga**:

* Github original: [(.csv)](https://raw.githubusercontent.com/ia4legos/MachineLearning/main/data/housing.csv)
* Github preprocesado: [(.csv)](https://raw.githubusercontent.com/ia4legos/MachineLearning/main/data/housingCA_clean.csv)

### Meat Spec

El departamento de calidad de una empresa de alimentación se encarga de medir el contenido en grasa de la carne que comercializa. Este estudio se realiza mediante técnicas de analítica química, un proceso relativamente costoso en tiempo y recursos. Una alternativa que permitiría reducir costes y optimizar tiempo es emplear un espectrofotómetro (instrumento capaz de detectar la absorbancia que tiene un material a diferentes tipos de luz en función de sus características) e inferir el contenido en grasa a partir de sus medidas. Por lo tanto, el objetivo que se persigue es predecir el contenido en grasa (`fat`) a partir de los valores dados por el espectrofotómetro.


**Características del banco de datos**:

* Variable respuesta: `fat`
* Tipo de variable respuesta: numérica
* Tipo de problema que se quiere resolver: regresión
* Valores perdidos: no
* Número de registros: 215
* Número de variables: 101


**Variables contenidas**:

* **Variables desde *V1* hasta *V100***: cada una de las variables hace referencia al valor de la muestra para una longitud de onda específica.
* **fat**: contenido en grasa de la carne (en gramos)


**Ficheros carga**:

* Github original: [(.csv)](https://raw.githubusercontent.com/ia4legos/MachineLearning/main/data/meatspec.csv)
* Github preprocesado: [(.csv)](https://raw.githubusercontent.com/ia4legos/MachineLearning/main/data/meatspec_clean.csv)

### Winequality

El conjunto de datos `Winequality` está relacionado con las variantes rojas del vino portugués "Vinho Verde". Debido a cuestiones de privacidad y logística, sólo se dispone de variables fisicoquímicas (entradas) y sensoriales (la salida). Por ejemplo, no hay datos sobre tipos de uva, marca de vino, precio de venta del vino, etc. Las variables provienen de tests psicoquímicos y el objetivo es, con las características disponibles, predecir la calidad del vino (`quality`) en la escala discreta de 0 a 10.


**Características del banco de datos**:
* Variable respuesta: `quality`
* Tipo de variable respuesta: numérica
* Tipo de problema que se quiere resolver: clasificación
* Valores perdidos: no
* Número de registros: 1599
* Número de variables: 12


**Variables contenidas**:
* **fixed acidity**: acidez fija.
* **volatile acidity**: acidez volátil.
* **citric acid**: acidez cítrica.
* **residual sugar**: azúcar residual (cantidad de azúcar que queda después de que se detiene la fermentación)
* **chlorides**: cloruros (cantidad de sal en el vino)
* **free sulfur dioxide**: dióxido de azufre libre.
* **total sulfur dioxide**: dióxido de azufre total.
* **density**: densidad.
* **pH**: ph.
* **sulphates**: sulfatos.
* **alcohol**: alcohol (porcentaje de contenido de alcohol del vino)
* **quality**: calidad (entre 0 y 10)

**Referencias bibliográficas**:

* P. Cortez, A. Cerdeira, F. Almeida, T. Matos and J. Reis.
Modeling wine preferences by data mining from physicochemical properties. In Decision Support Systems, Elsevier, 47(4):547-553, 2009.
* Kaggle: https://www.kaggle.com/datasets/sh6147782/winequalityred


**Ficheros carga**:

* Github original: [(.csv)](https://raw.githubusercontent.com/ia4legos/MachineLearning/main/data/winequality-red.csv)

### Student Alcohol Consumption (alcohol)

Los datos de consumo de alcohol entre adolescentes se obtuvieron en una encuesta realizada a los alumnos de los cursos de matemáticas y lengua portuguesa de enseñanza secundaria. Además de las variables referidas al consumo se dipone de mucha información social, de género y de estudio interesante sobre los estudiantes. Este banco de datos se puede utilizar para un análisis exploratorio de datos o para intentar predecir la nota final de los estudiantes.

Fte: https://www.kaggle.com/prashant111/student-alcohol-consumption


La información disponible es:

* **school** - student's school (binary: 'GP' - Gabriel Pereira or 'MS' - Mousinho da Silveira)
* **sex** - student's sex (binary: 'F' - female or 'M' - male)
* **age** - student's age (numeric: from 15 to 22)
* **address** - student's home address type (binary: 'U' - urban or 'R' - rural)
* **famsize** - family size (binary: 'LE3' - less or equal to 3 or 'GT3' - greater than 3)
* **Pstatus** - parent's cohabitation status (binary: 'T' - living together or 'A' - apart)
* **Medu** - mother's education (numeric: 0 - none, 1 - primary education (4th grade), 2 – 5th to 9th grade, 3 – secondary education or 4 – higher education)
* **Fedu** - father's education (numeric: 0 - none, 1 - primary education (4th grade), 2 – 5th to 9th grade, 3 – secondary education or 4 – higher education)
* **Mjob** - mother's job (nominal: 'teacher', 'health' care related, civil 'services' (e.g. administrative or police), 'at_home' or 'other')
* **Fjob** - father's job (nominal: 'teacher', 'health' care related, civil 'services' (e.g. administrative or police), 'at_home' or 'other')
* **reason** - reason to choose this school (nominal: close to 'home', school 'reputation', 'course' preference or 'other')
* **guardian** - student's guardian (nominal: 'mother', 'father' or 'other')
* **traveltime** - home to school travel time (numeric: 1 - 1 hour)
* **studytime** - weekly study time (numeric: 1 - 10 hours)
* **failures** - number of past class failures (numeric: n if 1<=n<3, else 4)
* **schoolsup** - extra educational support (binary: yes or no)
* **famsup** - family educational support (binary: yes or no)
* **paid** - extra paid classes within the course subject (Math or Portuguese) (binary: yes or no)
* **activities** - extra-curricular activities (binary: yes or no)
* **nursery** - attended nursery school (binary: yes or no)
* **higher** - wants to take higher education (binary: yes or no)
* **internet** - Internet access at home (binary: yes or no)
* **romantic** - with a romantic relationship (binary: yes or no)
* **famrel** - quality of family relationships (numeric: from 1 - very bad to 5 - excellent)
* **freetime** - free time after school (numeric: from 1 - very low to 5 - very high)
* **goout** - going out with friends (numeric: from 1 - very low to 5 - very high)
* **Dalc** - workday alcohol consumption (numeric: from 1 - very low to 5 - very high)
* **Walc** - weekend alcohol consumption (numeric: from 1 - very low to 5 - very high)
* **health** - current health status (numeric: from 1 - very bad to 5 - very good)
* **absences** - number of school absences (numeric: from 0 to 93)

In [ ]:
# LECTURA
url = 'https://raw.githubusercontent.com/jmsocuellamos/DataPython/main/student-mat.csv'
alcohol = pd.read_csv(url)
alcohol

### Crop yield

Este conjunto de datos contiene datos agrícolas de 1.000.000 de muestras destinadas a predecir el rendimiento de los cultivos (en toneladas por hectárea) en función de varios factores. El conjunto de datos se puede utilizar para predecir la productividad de los cultivos.

* `Region`: La región geográfica donde se cultiva el cultivo (Norte, Este, Sur, Oeste) (North, East, South, West).
* `Soil_Type`: El tipo de suelo en el que se planta el cultivo (Arcilloso, Arenoso, Franco, Limo, Turbio, Calcáreo) (Clay, Sandy, Loam, Silt, Peaty, Chalky).
* `Crop`: El tipo de cultivo cultivado (Trigo, Arroz, Maíz, Cebada, Soja, Algodón) ((Wheat, Rice, Maize, Barley, Soybean, Cotton)).
* `Rainfall_mm`: La cantidad de lluvia recibida en milímetros durante el período de crecimiento del cultivo.
* `Temperature_Celsius`: La temperatura promedio durante el período de crecimiento del cultivo, medida en grados Celsius.
* `Fertilizer_Used`: Indica si se aplicó fertilizante (True = Yes, False = No).
* `Irrigation_Used`: indica si se utilizó riego durante el período de crecimiento del cultivo (True = Yes, False = No).
* `Weather_Condition`: la condición climática predominante durante la temporada de crecimiento (soleado, lluvioso, nublado) (Sunny, Rainy, Cloudy).
* `Days_to_Harvest`: la cantidad de días que se necesitan para cosechar el cultivo después de la siembra.
* `Yield_tons_per_hectare` el rendimiento total del cultivo producido, medido en toneladas por hectárea.

Cargamos los datos originales y obtenemos la información correspondiente. Como los datos están disponibles en kaggle podemos aprovechar su api de descarga para acceder a los datos.

In [ ]:
# Cargamos el módulo para importar datos de kaggle
import kagglehub

# Obtenemos la dirección de descarga de la última versión del fichero en nuestro colab
path = kagglehub.dataset_download("samuelotiattakorah/agriculture-crop-yield")
%cd $path
# Cargamos el banco de datos
crop_yield = pd.read_csv(os.listdir(path)[0])
crop_yield.head(5)

### Southern California Energy Consumption

This dataset provides comprehensive energy consumption data collected from various buildings in Southern California, covering the period from January 2018 to January 2024. The data includes hourly electricity usage records for residential, commercial, and industrial buildings, along with environmental and operational metrics. The dataset is intended for researchers and practitioners interested in electricity consumption forecasting, energy management, sustainability, and the development of AI-based optimization models. The diverse features and real-world scenarios captured in the dataset make it suitable for time-series analysis, regression tasks, and classification problems.

The data was collected from over 100 facilities across Southern California, integrating information from smart meters, IoT sensors, building management systems, and regional utility companies. It covers different seasons, significant events (e.g., public holidays, extreme weather), and varying energy consumption patterns, allowing for robust analysis of electricity usage trends and energy-saving opportunities.

Dataset Features:

* `Timestamp` (DateTime): The date and time when the data was recorded, sampled hourly from 2018 to 2024.
* `Building Type` (Categorical): Type of building (Residential, Commercial, or Industrial).
* `Energy Consumption (kWh)` (Float): Total electricity consumption recorded in kilowatt-hours.
* `Temperature (°C)` (Float): Outdoor temperature at the time of consumption.
* `Humidity (%)` (Float): Relative humidity levels during the time of energy consumption.
* `Occupancy Rate (%)` (Float): Estimated percentage of occupied spaces in the building.
* `Lighting Consumption (kWh)` (Float): Electricity usage specifically for lighting within the building.
* `HVAC Consumption (kWh)` (Float): Energy used by heating, ventilation, and air conditioning systems.
* `Energy Price ($/kWh)` (Float): Cost of electricity at the time of consumption.
* `Carbon Emission Rate (g CO2/kWh)` (Float): Associated carbon emissions per unit of energy consumption.
* `Power Factor` (Float): Ratio of the real power used to the apparent power in the building.
* `Reactive Power (kVAR)` (Float): The reactive component of the electrical power consumed.
* `Wind Speed (m/s)` (Float): Wind speed at the time of the energy recording.
* `Solar Radiation (W/m²)` (Float): Solar energy incident on the building's location.
* `Rainfall (mm)` (Float): Recorded rainfall during the hour.
* `Thermal Comfort Index` (Float): An index representing perceived indoor comfort based on temperature and humidity.
* `Peak Demand Indicator` (Binary): Indicates whether the energy demand was during peak hours (1) or off-peak hours (0).
* `Energy Savings Potential (%)` (Float): Predicted potential savings in energy consumption based on current usage patterns.
* `Carbon Emission Reduction Category` (Categorical): Level of reduction in carbon emissions (Low, Medium, High, Very High).

In [ ]:
# Cargamos el módulo para importar datos de kaggle
import kagglehub

# Obtenemos la dirección de descarga de la última versión del fichero en nuestro colab
path = kagglehub.dataset_download("datasetengineer/southern-california-energy-consumption")
%cd $path
# Cargamos el banco de datos
california_electricity = pd.read_csv(os.listdir(path)[0])
california_electricity.head(5)

### Housing

El conjunto de datos de viviendas de Boston (`Housing`) contiene información recopilada por el Servicio del Censo de los EE. UU. sobre viviendas en el área de Boston MA. Como objetivo que se persigue predecir el valor medio de las viviendas ocupadas por sus propietarios en miles de dólares (`MEDV`) a partir del conjunto de variables predictoras. Las variables contenidas en el banco de datos son:

* `CRIM`: tasa de criminalidad per cápita por ciudad.
* `ZN`: proporción de terreno residencial zonificado para lotes de más de 25,000 pies cuadrados.
* `INDUS`: proporción de acres comerciales no minoristas por ciudad.
* `CHAS`: variable ficticia del río Charles (1 si el tramo limita con el río; 0 en caso contrario)
* `NOX`: concentración de óxidos nítricos (partes por 10 millones)
* `RM`: número promedio de cuartos por vivienda.
* `AGE`: proporción de unidades ocupadas por propietarios construidas antes de 1940.
* `DIS`: distancias ponderadas a cinco centros de empleo de Boston.
* `RAD`: índice de accesibilidad a las carreteras radiales.
* `TAX`: tasa de impuestos sobre la propiedad de valor total por cada $10,000.
* `PTRATIO`: ratio alumno-profesor por localidad.
* `B`: 1000(Bk - 0.63)^2 donde Bk es la proporción de negros por ciudad.
* `LSTAT`: % posición más baja de la población.

Cargamos los datos originales y obtenemos la información correspondiente:

In [ ]:
# Housing Dataset
url = 'https://raw.githubusercontent.com/ia4legos/MachineLearning/main/data/housing.data'
housing = pd.read_csv(url,header=None, sep='\s+')
# Nombres de las variables
names = ['CRIM', 'ZN', 'INDUS', 'CHAS','NOX', 'RM', 'AGE', 'DIS', 'RAD',
              'TAX', 'PTRATIO', 'B', 'LSTAT', 'MEDV']
# Asignamos nombres al banco de datos
housing.columns = names
housing.head(5)

## <font color="steelblue">Problemas de clasificación</font>

### Abalone

En este conjunto de datos se recoge información sobre los abulones (unos mejillones, de la familia de los moluscos). Se está interesado en medir su desarrollo, que viene determinado principalmente por su desarrollo sexual. Concretamente se consideran tres estados de desarrollo asociados con la variable `Sex`: `M` (machos), `F` (hembras), e `I` (infantil o sin desarrollo sexual). Para clasificar cada sujeto se utiliza el conjunto de variables que son de tipo numérico. La edad del abulón se determina cortando la cáscara a través del cono, tintándola y contando el número de anillos mediante un microscopio (una tarea tediosa que consume mucho tiempo).

**Características del banco de datos**:

* Variable respuesta: `Sex`
* Tipo de variable respuesta: categórica
* Tipo de problema que se quiere resolver: clasificación
* Valores perdidos: no
* Número de registros: 4177
* Número de variables: 8


**Variables contenidas**:
* **Sex**: sexo (M = masculino, F = femenino e I = intantil)
* **Length**: longitud de la carcasa más larga (en mm)
* **Diameter**: diámetro perpendicular a la longitud (en mm)
* **Height**: altura con carne en cáscara (en mm)
* **Whole weight**: peso completo (en gramos)
* **Shucked weight**: peso de la carne (en gramos)
* **Viscera weight**: peso de las vísceras después del sangrado (en gramos)
* **Shell weight**: peso de la cáscara después de secarse (en gramos)
* **Rings**: anillos (+1,5 da la edad en años)



**Referencias bibliográficas**:
* Kaggle: [enlace](https://www.kaggle.com/datasets/rodolfomendes/abalone-dataset)


**Ficheros carga**:

* Github original: [(.csv)](https://raw.githubusercontent.com/ia4legos/MachineLearning/main/data/abalone.csv)


### Breast Cancer Wisconsin

En esta base de datos se recoge información sobre los casos de cáncer de mama en la ciudad de Wisconsin durante un periodo determinado. Las características de la base de datos se calculan a partir de una imagen digitalizada de un aspiración de aguja fina (FNA) de una masa mamaria. Describen las características de los núcleos celulares presentes en la imagen y el objetivo que se persigue es clasificar un tumor como benigno o maligno en función de las características medidas sobre la imagen.


**Características del banco de datos**:

* Variable respuesta: `diagnosis`
* Tipo de variable respuesta: categórica (binaria)
* Tipo de problema que se quiere resolver: clasificación
* Valores perdidos: no
* Número de registros: 569
* Número de variables: 32


**Variables contenidas**:

* **id**: identificador.
* **diagnosis**: diagnóstico de tejidos mamarios (M = maligno, B = benigno)
* **radius_mean**: media de las distancias del centro a los puntos del perímetro.
* **texture_mean**: desviación estándar de los valores de la escala de grises.
* **perimeter_mean**:tamaño medio del tumor central.
* **area_mean**:
* **smoothness_mean**: media de variación local en longitudes de radio
* **compactness_mean**: (media de perímetro)^2 / área - 1,0
* **concavity_mean**: media de gravedad de las porciones cóncavas del contorno.
* **concave points_mean**: media para el número de porciones cóncavas del contorno.
* **symmetry_mean**:
* **fractal_dimension_mean**:
media para "aproximación de la costa" - 1.
* **radius_se**: error estándar para la media de distancias desde el centro hasta los puntos en el perímetro.
* **texture_se**: error estándar para la desviación estándar de los valores de escala de grises.
* **perimeter_se**:
* **area_se**:
* **smoothness_se**: error estándar para la variación local en las longitudes del radio.
* **compactness_se**: (error estándar para perímetro)^2 / área - 1,0.
* **concavity_se**: error estándar para la gravedad de las partes cóncavas del contorno.
* **concave points_se**: error estándar para el número de porciones cóncavas del contorno.
* **symmetry_se**:
* **fractal_dimension_se**: error estándar para "aproximación de la costa" - 1.
* **radius_worst**: "peor" o mayor valor medio para la media de distancias desde el centro hasta los puntos del perímetro (en cm).
* **texture_worst**: "peor" o mayor valor medio para la desviación estándar de los valores de escala de grises.
* **perimeter_worst**:
* **area_worst**:
* **smoothness_worst**: "peor" o mayor valor medio para la variación local en longitudes de radio.
* **compactness_worst**: "peor" o mayor valor medio para el perímetro^2 / área - 1,0.
* **concavity_worst**: "peor" o mayor valor medio para la gravedad de las porciones cóncavas del contorno.
* **concave points_worst**: "peor" o mayor valor medio para el número de porciones cóncavas del contorno.
* **symmetry_worst**:
* **fractal_dimension_worst**: "peor" o mayor valor medio para "aproximación de la costa" - 1.


**Referencias bibliográficas**:
* Kaggle: [enlace](https://www.kaggle.com/datasets/uciml/breast-cancer-wisconsin-data)


**Ficheros carga**:

* Github original: [(.csv)](https://raw.githubusercontent.com/ia4legos/MachineLearning/main/data/cancer.csv)
* Github preprocesado: [(.csv)](https://raw.githubusercontent.com/ia4legos/MachineLearning/main/data/cancer_clean.csv)

### Iris

La base de datos iris consta de 50 muestras de cada una de las tres especies de Iris, una planta con flores nativa del este de América del Norte: setosa,  virgínica y versicolor. Se escogieron cuatro características de cada muestra: la longitud y el ancho de los sépalos y pétalos, en centímetros. El objetivo que se persigue es predecir la especie (`species`) en función de las características de los pétalos y sépalos medidas en sus flores.

**Características del banco de datos**:

* Variable respuesta: `species`
* Tipo de variable respuesta: categórica
* Tipo de problema que se quiere resolver: clasificación
* Valores perdidos: no
* Número de registros: 150
* Número de variables: 5


**Variables contenidas**:
* **sepal_length**: longitud del sépalo (en cm)
* **sepal_width**: anchura del sépalo (en cm)
* **petal_length**: longitud del pétalo (en cm)
* **petal_width**: anchura del pétalo (en cm)
* **species**: especie de la flor (Iris Setosa, Iris virginica e Iris versicolor)


**Referencias bibliográficas**:
* Kaggle: [enlace](https://www.kaggle.com/datasets/uciml/iris)


**Ficheros carga**:

* Github original: [(.csv)](https://raw.githubusercontent.com/ia4legos/MachineLearning/main/data/IRIS.csv)
* Github preprocesado: [(.csv)](https://raw.githubusercontent.com/ia4legos/MachineLearning/main/data/iris_clean.csv)

### Stroke prediction dataset

Según la Organización Mundial de la Salud (OMS), el ictus es la segunda causa de muerte en el mundo, responsable de aproximadamente el 11% del total de fallecimientos. El banco de datos `Stroke` se utiliza para predecir la probabilidad de que un paciente sufra un ictus en función de sus características, como el sexo, la edad, diversas enfermedades y estatus de fumador. Cada fila de los datos proporciona información relevante sobre el paciente.

**Características del banco de datos**:
* Variable respuesta: `stroke`
* Tipo de variable respuesta: categórica
* Tipo de problema que se quiere resolver: clasificación
* Valores perdidos: sí (en la variable `bmi`)
* Número de registros: 5110
* Número de variables: 12

**Variables contenidas**:
* **id**: identificador único del sujeto
* **gender**: sexo ("Male" = masculino, "Female" = femenino y "Other" = otro)
* **age**: edad del paciente (en años)
* **hypertension**: el paciente tiene hipertensión ("Yes" = sí y "No")
* **heart_disease**:  el paciente tiene una enfermedad del corazón ("Yes" = sí y "No")
* **ever_married**:  el paciente se ha casado alguna vez ("Yes" = sí y "No")
* **work_type**: tipo de trabajo ("children" = niños, "Govt_jov" = Gubernamental, "Never_worked" = nunca ha trabajado, "Private" = privado y "Self-employed" = autónomo)
* **Residence_type**: tipo de residencia ("Rural" y "Urban" =  urbana)
* **avg_glucose_level**: media de glucosa en sangre.
* **bmi**: índice de masa corporal.
* **smoking_status**: nivel de fumador ("formerly smoked" = fumaba, "never smoked" = nunca ha fumado, "smokes" = fuma y "Unknown" = desconocido).
* **stroke**: el paciente sufre un ictus ("Yes" = sí y "No")

**Referencias bibliográficas**:

* kaggle: https://www.kaggle.com/datasets/fedesoriano/stroke-prediction-dataset

**Ficheros carga**:

* Github original: [(.csv)](https://raw.githubusercontent.com/ia4legos/MachineLearning/main/data/stroke_ori.csv)

* Github preprocesado: [(.csv)](https://raw.githubusercontent.com/ia4legos/MachineLearning/main/data/stroke_clean.csv)

### Vehicle Silhouettes

Este conjunto de datos recoge información de cuatro tipos diferentes de vehículos, utilizando un conjunto de imágenes, y en particular de características extraídas de su silueta. Para el experimento se utilizaron cuatro vehículos modelo “Corgie”: un bus de dos pisos, una camioneta Cheverolet, un Saab 9000 y un Opel Manta 400. El objetivo del estudio las siluetas como relativa a alguno de los cuatro tipos de vehículos. Todos los atributos son numéricos discretos salvo la primera variable que lleva el registro de la observación, y muchos de ellos contienen valores perdidos.

**Características del banco de datos**:

* Variable objetivo: `class`
* Tipo de variable objetivo: categórica
* Tipo de problema que se quiere resolver: clasificación
* Valores perdidos: no
* Número de registros: 946
* Número de variables: 19

**Variables contenidas**:

* **compactness**: se calcula mediante la fórmula $(perímetro.medio)^2/área$
* **circularity**: se calcula mediante la fórmula $(radio. medio)^2/área$
* **distance_circularity**: se calcula mediante la fórmula $área/(distancia. media. desde. el. borde)^2$
* **radius_ratio**: se calcula mediante la fórmula $(max.rad-min.rad)/av.radius$
* **pr.axis_aspect_ratio**: se calcula mediante la fórmula $(eje. menor)/(eje. mayor)$
* **max.length_aspect_ratio**: se calcula mediante la fórmula $(longitud. perp. longitud. máxima)/(longitud. máxima)$
* **scatter_ratio**: se calcula mediante la fórmula $ (inercia. sobre. el. eje. menor)/(inercia. sobre. el. eje. mayor)$
* **elongatedness**: se calcula mediante la fórmula $área/(ancho. de. contracción)^2$
* **pr.axis_rectangularity**: se calcula mediante la fórmula $área/(longitud. del. eje. pr. * ancho. del. eje. pr.)$
* **max.length_rectangularity**: se calcula mediante la fórmula $área/(long. max. * long. perp.)$
* **scaled_variance**: se calcula mediante la fórmula $(momento. de. segundo. orden. sobre. el. eje. menor)/área. a. lo. largo. del. eje. mayor$
* **scaled_variance.1**: se calcula mediante la fórmula $(momento. de. segundo. orden. sobre. el. eje. mayor)/área. a. lo. largo. del. eje. menor$
* **scaled_radius_of_gyration**: se calcula mediante la fórmula $(mavar+mivar)/área$
* **scaled_radius_of_gyration.1**: se calcula mediante la fórmula $(mavar+mivar)/área$
* **skewness_about**: se calcula mediante la fórmula $(momento. de. tercer. orden. sobre. el. eje. principal.)/sigma-min^3 * eje. principal$

* **skewness_about.1**: se calcula mediante la fórmula (momento. de. tercer. orden. sobre. el. eje. menor.)/sigma-maj^3 * eje. menor$

* **skewness_about.2**: se calcula mediante la fórmula (momento. de. tercer. orden. sobre. el. eje. menor.)/sigma-maj^3 * eje. menor$

* **hollows_ratio**: se calcula mediante la fórmula $(área. de. huecos)/(área. del. polígono. delimitador)$
* **class**: tipo de vehículo (van, saab, bus y opel)


**Referencias bibliográficas**:
* Turing Institute Research Memorandum TIRM-87-018 "Vehicle Recognition Using Rule Based Methods" by Siebert,JP (March 1987)
[Web Link].


**Ficheros carga**:
* Github original: [(.csv)]('https://raw.githubusercontent.com/ia4legos/MachineLearning/main/data/vehicle.csv')

### Water Potability

El agua potable es el derecho humano más básico y un factor importante para la salud, e indicadores relacionados con el desarrollo a nivel nacional, regional y local. En algunas regiones, se ha demostrado que las inversiones en abastecimiento de agua y saneamiento pueden producir un beneficio económico neto, ya que la reducción de los efectos adversos para la salud y los costes de la atención sanitaria superan los costes de las intervenciones. El conjunto de datos `Water potability`, tiene por objetivo estudiar la potabilidad del agua utilizando varias propiedades químicas.

**Características del banco de datos**:

* Variable respuesta: `potability`
* Tipo de variable respuesta: categórica
* Tipo de problema que se quiere resolver: clasificación
* Valores perdidos: sí (variables `ph`, `Sulfate` y `Trihalomethanes`)
* Número de registros: 3276
* Número de variables: 10


**Variables contenidas**:
* **pH**: valor del pH.
* **Hardness**: dureza o capacidad del agua para precipitar el jabón causado por el calcio y el magnesio.
* **Solids**: sólidos disueltos totales (en partes por millón)
* **Chloramines**: cantidad de cloraminas (en partes por millón)
* **Sulfate**: cantidad de sulfatos disueltos (en mg/L)
* **Conductivity**: conductividad eléctrica del agua (en μS/cm)
* **Organic_carbon**: cantidad de carbono orgánico (en partes por millón)
* **Trihalomethanes**: cantidad de trihalometanos (en μg/L)
* **Turbidity**: medida de la propiedad de emisión de luz del agua en NTU.
* **Potability**: indica si el agua es segura para el consumo humano (1 = potable y 0 = no potable)



**Referencias bibliográficas**:

* Kaggle: https://www.kaggle.com/datasets/adityakadiwal/water-potability


**Ficheros carga**:

* Github original: [(.csv)](https://raw.githubusercontent.com/ia4legos/MachineLearning/main/data/water_potability.csv)
* Github preprocesado: [(.csv)](https://raw.githubusercontent.com/ia4legos/MachineLearning/main/data/water_clean.csv)

### Wine recognition

En este banco de datos se recogen los resultados del análisis químico de vinos cultivados en la misma región de Italia, pero procedentes de tres tipos de cultivo distintos. El análisis determinó las cantidades de 13 características que se encuentran en cada una de las muestras de vinos. El objetivo que perseguimos es clasificar cada muestra en una de estas tres clases de vino (`Class label`) en función de sus características químicas.


**Características del banco de datos**:
* Variable respuesta: `Class label`
* Tipo de variable respuesta: categórica
* Tipo de problema que se quiere resolver: clasificación
* Valores perdidos: no
* Número de registros: 178
* Número de variables: 14


**Variables contenidas**:

* **Class label**: etiqueta que indica el tipo de vino (1, 2 o 3)
* **Alcohol**: alcohol
* **Malic acid**: ácido málico
* **Ash**: ceniza
* **Alcalinity of ash**: alcalinidad de la ceniza
* **Magnesium**: magnesio
* **Total phenols**: fenoles totales
* **Flavanoids**: flavonoides
* **Nonflavanoid phenols**: fenoles no flavonoides
* **Proanthocyanins**: proantocianinas
* **Color intensity**: intensidad de color
* **Hue**: matiz
* **OD280/OD315 of diluted wines**: OD280/OD315 de vinos diluidos
* **Proline**: prolina


**Referencias bibliográficas**:
* Kaggle: [enlace](https://www.kaggle.com/datasets/amirhosseinzinati/wine-recognition-data)


**Ficheros carga**:

* Github original: [(.csv)](https://archive.ics.uci.edu/ml/machine-learning-databases/wine/wine.data)
* Github preprocesado: [(.csv)](https://raw.githubusercontent.com/ia4legos/MachineLearning/main/data/vinos_clean.csv)

### Pima Indian Diabetes (diabetes)

Este conjunto de datos procede del Instituto Nacional de Diabetes y Enfermedades Digestivas y Renales. El objetivo del estudio es predecir si un paciente tiene o no diabetes, basándose en determinadas mediciones diagnósticas incluidas en el conjunto de datos. Los datos considerados provienen de imponer una serie de restricciones a una base de datos más extensa; en concreto, todos los pacientes son mujeres de al menos 21 años de edad y de ascendencia india pima.
Los datos contienen diversas variables médicas predictoras y una variable objetivo, *Outcome*, que indica si la paciente tiene o no diabetes. Las variables predictoras incluyen el número de embarazos que ha tenido la paciente, su IMC, su nivel de insulina, su edad, etc.

Fte: https://www.kaggle.com/uciml/pima-indians-diabetes-database


La información disponible es:

* **Pregnancies**: Number of times pregnant
* **Glucose**: Plasma glucose concentration a 2 hours in an oral glucose tolerance test
* **BloodPressure**: Diastolic blood pressure (mm Hg)
* **SkinThickness**: Triceps skin fold thickness (mm)
* **Insulin**: 2-Hour serum insulin (mu U/ml)
* **BMI**: Body mass index (weight in kg/(height in m)^2)
* **DiabetesPedigreeFunction**: Diabetes pedigree function
* **Age**: Age (years)
* **Outcome**: Class variable (0 or 1) 268 of 768 are 1, the others are 0

In [ ]:
# LECTURA
url = 'https://raw.githubusercontent.com/jmsocuellamos/DataPython/main/diabetes.csv'
diabetes = pd.read_csv(url)
diabetes

### Clasificación de pasas

Se obtuvieron imágenes de las variedades de pasas Kecimen y Besni cultivadas en Turquía con CVS. Se utilizaron un total de 900 granos de pasas, incluidas 450 piezas de ambas variedades. Estas imágenes se sometieron a varias etapas de preprocesamiento y se extrajeron 7 características morfológicas. El objetivo es clasificar cada grano de pasa en función de las predictoras morfólogicas extraídas a través de las imágenes obtenidas.

Las variables consideradas son:
* `Area`: Proporciona el número de píxeles dentro de los límites de la imagen para cada pasa.
* `MajorAxisLength`: Da la longitud del eje principal, que es la línea más larga que se puede dibujar en la imagen de cada pasa.
* `MinorAxisLength`: Da la longitud del eje secundario, que es la línea más corta que se puede dibujar en la imagen de cada pasa.
* `Eccentricity`: Da una medida de la excentricidad de la elipse, que tiene los mismos momentos que las pasas.
* `ConvexArea`: Da el número de píxeles de la capa convexa más pequeña de la región formada por la pasa.
* `Extent`: Proporciona la relación entre la región formada por la pasa y el total de píxeles en el cuadro delimitador.
* `Perimeter`: Mide el entorno calculando la distancia entre los límites de la pasa y los píxeles que la rodean.
* `Class`: identificador del tipo de pasa Kecimen o Besni.

Cargamos el banco de datos desde la plataforma kaggle.

In [ ]:
# Cargamos el módulo para importar datos de kaggle
import kagglehub

# Obtenemos la dirección de descarga de la última versión del fichero en nuestro colab
path = kagglehub.dataset_download("nimapourmoradi/raisin-binary-classification")
%cd $path
# Cargamos el banco de datos
raisins = pd.read_csv("Raisin_Dataset.csv")
raisins.head(5)

### Admisión a MBA

El conjunto de datos se refiere a los resulatdos de admisión al rpgrama MBA de la universidad de Warton para el año 2025. las variables recogidas son:

* `application_id`: identificador de la solicitud de admisión.
* `gender`: sexo del solicitante (Male, Female)
* `international`: se trata de un estudiante internacional (TRUE/FALSE)
* `gpa`: promedio de calificaciones del solicitante en la escala de 1 a 4
* `major`: área de la titulación universitaria (Business, STEM, Humanities)
* `race`: raza del solicitante (White, Black, Asian, Hispanic, Other / null: international student)
* `gmat`: puntuación del GMAT (examen de acceso de matemáticas) del solicitante (máxima puntuación  = 800)
* `work_exp`: Número de años de experiencia laboral
* `work_industry`: Industria en la que el solicitante tiene experiencia laboral previa (Consulting, Finance, Technology, etc.)
* `admission`: Resultado de la admisión (Admit, Waitlist, Null: Deny)

El objetivo es crear un modelo que nos permita automatizar el proceso de admisión. Este conjunto de datos tiene la carcaterística especial de que los valores pérdidos lo son tales sino que no se ha completado dicho campo. una de nuestras primeras tareas al cargar los datos será modificar dichos valores.

Cargamos los datos desde el repositorio kaggle

In [ ]:
# Cargamos el módulo para importar datos de kaggle
import kagglehub

# Obtenemos la dirección de descarga de la última versión del fichero en nuestro colab
path = kagglehub.dataset_download("taweilo/mba-admission-dataset")
%cd $path
# Cargamos el banco de datos
admision_mba = pd.read_csv(os.listdir(path)[0])
admision_mba.head(5)

### Score crediticio

En una emepresa crediticia están tratando de desarrollar una herramienta automática que permita establecer el riesgo de conceder un préstamo en función de las características del solicitante. Durante un año, la empresa ha recopilado datos bancarios básicos y ha reunido mucha información relacionada con el crédito. La gerencia quiere construir un sistema inteligente para separar a las personas en categorías de puntaje crediticio para reducir los esfuerzos manuales. Las varaibles recogidas para cada solicitante son:

* `Age`: Edad del cliente
* `Occupation`: Ocupación del cliente
* `Annual_Income`: Ingreso anual del cliente
* `Num_Bank_Accounts`: Número de cuentas bancarias que posee el cliente
* `Num_Credit_Card`: Número de tarjetas de crédito que posee el cliente
* `Interest_Rate`: Tasa de interés de la tarjeta de crédito
* `Num_of_Loan`: Número de préstamos tomados
* `Delay_from_due_date`: Número promedio de días de retraso desde la fecha de vencimiento del pago
* `Num_of_Delayed_Payment`: Número de pagos retrasados
* `Changed_Credit_Limit`: Cambio porcentual en el límite de la tarjeta de crédito
* `Num_Credit_Inquiries`: Número de consultas sobre tarjetas de crédito realizadas por los prestamistas al cliente
* `Credit_Mix`: Tipo de combinación de créditos que tiene el cliente
* `Outstanding_Debt`: Deuda restante por pagar (en USD)
* `Credit_Utilization_Ratio`: ratio de utilización de la tarjeta de crédito
* `Credit_History_Age`: antigüedad del historial crediticio del cliente en meses
* `Payment_of_Min_Amount`: si el cliente pagó o no el importe mínimo
* `Total_EMI_per_month`: pagos mensuales de EMI (en USD)
* `Amount_invested_monthly`: importe mensual invertido por el cliente (en USD)
* `Payment_Behaviour`: comportamiento de pago del cliente
* `Monthly_Balance`: importe del saldo mensual del cliente (en USD)
* `Credit_Score`: Puntaje crediticio asignado al cliente para el mes determinado
* `Credit_History_Age_Months`

Cargamos los datos y vemos los tipos de las variables.

In [ ]:
# Cargamos el módulo para importar datos de kaggle
import kagglehub

# Obtenemos la dirección de descarga de la última versión del fichero en nuestro colab
path = kagglehub.dataset_download("eneskosar19/credit-score-classification-clean-data")
%cd $path
# Cargamos el banco de datos
credit_score = pd.read_csv(os.listdir(path)[0])
credit_score.head(5)